File: leetcode/cpp/40. 组合总和 II.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& ret, vector<int>& path, const vector<int>& nums, int target, int i) {
        if (target == 0) {
            ret.push_back(path);
        } else if (target > 0 && i < nums.size()) {
            for (int j = i; j < nums.size(); j++) {
                if (find(nums.begin() + i, nums.begin() + j, nums[j]) == nums.begin() + j) {
                    path.push_back(nums[j]);
                    dfs(ret, path, nums, target - nums[j], j + 1);
                    path.pop_back();
                }
            }
        }
        return;
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> path;
        vector<vector<int>> ret;
        if (!candidates.empty()) {
            sort(candidates.begin(), candidates.end());
            dfs(ret, path, candidates, target, 0);
        }
        return ret;
    }
};

File: leetcode/python/40. 组合总和 II.py
#!/bin/bash python
"""
40. 组合总和 II

思路：
正常DFS
"""

class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        def dfs(ret, path, nums, target, i):
            """
            每个元素只有选或者不选两种情况，其实就是遍历所有子集，然后保证求和
            """
            if target == 0:
                ret.append(path[:])
            elif target > 0 and i < len(nums):
                for j in range(i, len(nums)):
                    if nums[j] in nums[i:j]:
                        continue
                    path.append(nums[j])
                    dfs(ret, path, nums, target - nums[j], j + 1)
                    path.pop()
            return
        # ===== 正式调用 =====
        ret, path = [], []
        if candidates != []:
            candidates.sort()
            dfs(ret, path, candidates, target, 0)
        return ret